### PR TITLE
8324514: ClassLoaderData::print_on should print address of class loader

### PR DESCRIPTION
**Pull request by shipilev**
This looks like a simple diagnostic regression from [JDK-8201556](https://bugs.openjdk.org/browse/JDK-8201556). Instead of printing the address of the `OopHandle` slot, we really want to print the address of the classloader that handle holds. That was not intentional, right, @coleenp?

I think we can `peek()` into handle, but the whole reason for [JDK-8201556](https://bugs.openjdk.org/browse/JDK-8201556) seems to avoid touching the classloader oops when unloading, and `peek()` would do a GC barrier on it. We could specialize for `!unloading` path, but I think we still want to see what is in the handle. So, this PR introduces the "peek_raw" version that would compile straight to plain access, after the null check.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324514](https://bugs.openjdk.org/browse/JDK-8324514): ClassLoaderData::print_on should print address of class loader (**Enhancement** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17534/head:pull/17534` \
`$ git checkout pull/17534`

Update a local copy of the PR: \
`$ git checkout pull/17534` \
`$ git pull https://git.openjdk.org/jdk.git pull/17534/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17534`

View PR using the GUI difftool: \
`$ git pr show -t 17534`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17534.diff">https://git.openjdk.org/jdk/pull/17534.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17534#issuecomment-1905883561)

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 1]
```diff
@@ -1003,7 +1003,7 @@ void ClassLoaderData::print_on(outputStream* out) const {
     _holder.print_on(out);
     out->print_cr("");
   }
-  out->print_cr(" - class loader        " INTPTR_FORMAT, p2i(_class_loader.ptr_raw()));
+  out->print_cr(" - class loader        " INTPTR_FORMAT, p2i(_class_loader.peek_raw()));
```
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/classfile/classLoaderData.cpp:1006]: Yes, I suppose we did want to print the oop not the OopHandle pointer.

[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/classfile/classLoaderData.cpp:1006]: If not peek(), I kind of think I'd prefer the !_unloading test more than a peek_raw().

[the remark this answers is not in the document]
Comment (kimbarrett) [reply src/hotspot/share/classfile/classLoaderData.cpp:1006]: I haven't looked at this change at all, and I can already say I completely agree with @coleenp here.

[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/classfile/classLoaderData.cpp]: Well, yes, it is dangerous in the sense that you cannot dereference the `oop` safely after this. It is only useful for debugging printouts. We can do `!unloading` too, like the very first commit.

**Review by kimbarrett — APPROVED** [review]
Looks good.

**Review by coleenp — APPROVED** [review]
Yes, this looks really good, plus with the <unloading, oop is bad> message describes why this code is special.  Thank you!

**Comment by shipilev** [comment]
Thanks!

/integrate

**Pull request closed without merging**